tools/SwiftSyntax/Diagnostic.swift

```diff
@@ -0,0 +1,159 @@
+//===--------------- Diagnostics.swift - Diagnostic Emitter ---------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+// This file provides the Diagnostic, Note, and FixIt types.
+//===----------------------------------------------------------------------===//
+
+/// Represents a source location in a Swift file.
+public struct SourceLocation {
+  /// The line in the file where this location resides.
+  public let line: Int
+
+  /// The UTF-8 byte offset from the beginning of the line where this location
+  /// resides.
+  public let column: Int
+
+  /// The UTF-8 byte offset into the file where this location resides.
+  public let offset: Int
+
+  /// The file in which this location resides.
+  public let file: String
+}
+
+/// Represents a start and end location in a Swift file.
+public struct SourceRange {
+  /// The beginning location in the source range.
+  public let start: SourceLocation
+
+  /// The ending location in the source range.
+  public let end: SourceLocation
```

Comment:
Do we ever need ending line/column/file information for ranges, or would a "length" field suffice?

Reply:
Would we lose multi-line ranges?

Reply:
Granted, clang and Swift don't really handle multi-line ranges very well (at all?)

Reply:
I assumed it would be length-in-bytes, which is fine for a fix-it and probably sufficient for a highlight.

Reply:
Clang and Swift both handle multi-line ranges fine as long as you're not using textual diagnostic output.

Reply:
Seems we do need an end location, at least for highlights in the `.dia` format.